src/runner.ts

```diff
@@ -0,0 +1,33 @@
+import { validate } from "@eng-automation/js";
+import { parse } from "yaml";
+
+import { Inputs } from ".";
+import { ConfigurationFile } from "./file/types";
+import { schema } from "./file/validator";
+import { PullRequestApi } from "./github/pullRequest";
+import { ActionLogger } from "./github/types";
+
+/** Action in charge of running the GitHub action */
+export class ActionRunner {
+  constructor(private readonly prApi: PullRequestApi, private readonly logger: ActionLogger) {}
+
+  /**
+   * Fetches the configuration file, parses it and validates it.
+   * If the config is invalid or not found, an error will be thrown.
+   */
+  async getConfigFile(configLocation: string): Promise<ConfigurationFile> {
+    const content = await this.prApi.getConfigFile(configLocation);
+    this.logger.debug(content);
+    const config: unknown = parse(content);
+
+    this.logger.info(`Obtained config at ${configLocation}`);
+
+    return validate<ConfigurationFile>(config, schema, { message: "Configuration file is invalid" });
+  }
+
+  async runAction(inputs: Omit<Inputs, "repoToken">): Promise<boolean> {
+    const config = await this.getConfigFile(inputs.configLocation);
+
+    return config !== null;
```

Comment:
This expression will always be true, as `getConfigFile` would throw the error, instead of returning `null`.  
However, I'm generally against returning `boolean` as an indicator of success, and would prefer this to be `Promise<void>`, with throwing on errors (which it does already). 
Boolean status says nothing about cause of the error. 

This way, you would need to change `src/index.ts` like this:

```ts
runner
  .runAction(inputs)
  .then(() => info("Action completed succesfully"))
  .catch(error => setFailed(error.message));
```

Reply:
> This expression will always be true

That is intended as there is no evaluation yet.

> However, I'm generally against returning `boolean` as an indicator of success, and would prefer this to be `Promise<void>`, with throwing on errors (which it does already). 

I have a reasoning for this:
- Error: Something unexpected went wrong (invalid config, exception, etc).
- false boolean: The action actually finished but the evaluation failed.
- positive boolean: The action succeed.

Because of the complexity of this action, we need to evaluate that the rules are being correctly executed, and the users are assigned, so we can not stop when a rule isn't fulfilled. So we would have to finish the whole function, and then run a simple evaluation to see if we throw an error or not.

```typescript
async runAction(inputs: any): Promise<void> {
  // thing 1
  // thing 2
  // thing 3 which evaluates

  // if the evaluation fails, we still need to move forward and can not stop here
  const errors = evaluateRules(a, b, c);

  // thing 4 which does more parsing
  // thing 5 which assigns missing reviewers
  // thing 6 which writes a report
  
  // now we could return errors.length === 0, or we throw an Error
  if (errors.length > 0) {
    throw new Error("Missing reviews");
  }
}
```

Also, [throwing exceptions as a control flow is usually considered an anti-pattern](https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why) (and I have seen how this anti-pattern did some significant damage in other projects)

> It has been said a thousand times, that exceptions are meant for exceptional conditions. A broken database connection is an exceptional condition. A user entering letters in an input field that should only allow numbers is not.

Reply:
> Also, [throwing exceptions as a control flow is usually considered an anti-pattern](https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why) (and I have seen how this anti-pattern did some significant damage in other projects)

Are we going back to `Ok | Err` pattern then?

Reply:
Even if you want to differentiate between errors and failed checks, returning boolean is a terrible way to do that. 
Failed checks should provide context which would be reported to the user. 

But I couldn't guess (or, failed at it) your intentions from your PR, as you didn't state what you indended to do. All I see is parsing config, which shouldn't fail unless there's an exceptional condition.  
Adding TODOs would help.

Reply:
[Done](https://github.com/paritytech/review-bot/issues/18) 😄